external/llvm-project/mlir/include/mlir/IR/AttrTypeBase.td

```diff
@@ -262,6 +262,60 @@ class AttrDef<Dialect dialect, string name, list<Trait> traits = [],
                                  "::" # cppClassName # ">($_self)">;
 }
 
+// Define an Attr name for NamedAttribute's.
+// - Name is dialect-scoped when not-inherent.
+// - Utilities to has/get/set/lookup on an Operation
+class AttrNameDef<Dialect dialect, string name, string userName,
+    string valueAttrType = "::mlir::Attribute"> :
+  AttrDef<dialect, name, [], "::mlir::StringAttr"> {
+  let mnemonic = userName;
+
+  string qualName = dialect.name # "." # mnemonic;
+  code typedefValueAttr = "typedef " # valueAttrType # " ValueAttrType;\n";
+  code getNameFunc = "static constexpr llvm::StringLiteral getName() { return \""
+      # qualName # "\"; }\n";
+
+  code nameFuncs = !strconcat(typedefValueAttr, getNameFunc, [{
+    // Get name based on inherentness
+    static llvm::StringLiteral getName(Operation *op) {
+      auto mnemonic = getMnemonic();
+      return op == nullptr || op->getInherentAttr(mnemonic) ? mnemonic : getName();
+    }
+    // Getters
+    static bool hasOn(::mlir::Operation *op) {
+      return op->hasAttr(getName(op));
+    }
+    static ValueAttrType getOn(::mlir::Operation *op) {
+      return op->getAttrOfType<ValueAttrType>(getName(op));
+    }
+    static ValueAttrType lookupOn(::mlir::Operation *op) {
+      if (auto attr = getOn(op))
+        return attr;
+      std::optional<RegisteredOperationName> opInfo = op->getRegisteredInfo();
```

Comment:
Strong object to this logic under this name.

Perhaps `findInContext` to have this recursive behavior

Reply:
Is it a **strong object** to the name or the logic?

Reply:
The logic itself is something that could exist. It's a strong object to it existing under this name in this context.

`lookupOn(Operation *op)` should not be doing a tree-walk.

`findNearestTo(Operation *op)` or a similar name should do the traversal you have in `lookupOn(Operation *op)`

Reply:
This is a hierarchical scope lookup based on the philosophy of inheritance. Operations under a loop, function or module can inherit properties based on their `parent` scope. I don't think findNearestTo captures this idea.